substrate/client/transaction-pool/src/fork_aware_txpool/revalidation_worker.rs

```diff
@@ -0,0 +1,264 @@
+		}
+	}
+}
+
+/// A Revalidation queue.
+///
+/// Allows to send the revalidation requests to the [`RevalidationWorker`].
+pub struct RevalidationQueue<Api, Block>
+where
+	Api: ChainApi<Block = Block> + 'static,
+	Block: BlockT,
+{
+	background: Option<TracingUnboundedSender<WorkerPayload<Api, Block>>>,
+}
+
+impl<Api, Block> RevalidationQueue<Api, Block>
+where
+	Api: ChainApi<Block = Block> + 'static,
+	Block: BlockT,
+	<Block as BlockT>::Hash: Unpin,
+{
+	/// New revalidation queue without background worker.
+	///
+	/// All validation requests will be blocking.
+	pub fn new() -> Self {
+		Self { background: None }
+	}
+
+	/// New revalidation queue with background worker.
+	///
+	/// All validation requests will be executed in the background.
+	pub fn new_with_worker() -> (Self, Pin<Box<dyn Future<Output = ()> + Send>>) {
+		let (to_worker, from_queue) = tracing_unbounded("mpsc_revalidation_queue", 100_000);
+		(Self { background: Some(to_worker) }, RevalidationWorker::new().run(from_queue).boxed())
+	}
+
+	/// Queue the view for later revalidation.
+	///
+	/// If the queue is configured with background worker, this will return immediately.
+	/// If the queue is configured without background worker, this will resolve after
+	/// revalidation is actually done.
+	///
+	/// Schedules execution of the [`View::revalidate`].
+	pub async fn revalidate_view(
+		&self,
+		view: Arc<View<Api>>,
+		finish_revalidation_worker_channels: FinishRevalidationWorkerChannels<Api>,
+	) {
+		log::trace!(
+			target: LOG_TARGET,
+			"revalidation_queue::revalidate_view: Sending view to revalidation queue at {}",
+			view.at.hash
+		);
+
+		if let Some(ref to_worker) = self.background {
+			if let Err(e) = to_worker.unbounded_send(WorkerPayload::RevalidateView(
+				view,
+				finish_revalidation_worker_channels,
+			)) {
+				log::warn!(target: LOG_TARGET, "revalidation_queue::revalidate_view: Failed to update background worker: {:?}", e);
+			}
+		} else {
+			view.revalidate(finish_revalidation_worker_channels).await
+		}
+	}
+
+	/// Revalidates the given mempool instance.
+	///
+	/// If queue configured with background worker, this will return immediately.
+	/// If queue configured without background worker, this will resolve after
+	/// revalidation is actually done.
+	///
+	/// Schedules execution of the [`TxMemPool::revalidate`].
+	pub async fn revalidate_mempool(
+		&self,
+		mempool: Arc<TxMemPool<Api, Block>>,
+		finalized_hash: HashAndNumber<Block>,
+	) {
+		log::trace!(
+			target: LOG_TARGET,
+			"Sent mempool to revalidation queue at hash: {:?}",
+			finalized_hash
+		);
+
+		if let Some(ref to_worker) = self.background {
+			if let Err(e) =
+				to_worker.unbounded_send(WorkerPayload::RevalidateMempool(mempool, finalized_hash))
+			{
+				log::warn!(target: LOG_TARGET, "Failed to update background worker: {:?}", e);
+			}
+		} else {
+			mempool.revalidate(finalized_hash).await
+		}
+	}
+}
+
+#[cfg(test)]
+//todo: add tests [#5480]
+mod tests {
```

Comment:
I am in favor of removing this and keeping the todo with issue number.

Reply:
test enabled in 571d598d901